cub/docs/developer_overview.rst

```diff
@@ -744,3 +744,198 @@ contains hidden visibility attribute.
+   reduce::sum(begin, end); 
+
+   // run-to-run determinism is enforced
+   reduce::sum(begin, end, cub::require(cub::run_to_run_determinism)); 
+
+   // default determinism guarantees, memory footprint limit
+   reduce::sum(begin, end, cub::require(cub::max_memory_footprint(128_KB)));
+
+   // default determinism guarantees, memory footprint limit
+   reduce::sum(begin, end, cub::require(cub::max_memory_footprint(128_KB)));
+
+   // run-to-run deterministic with memory footprint limit
+   reduce::sum(begin, end, cub::require(cub::run_to_run_determinism, 
+                                        cub::max_memory_footprint(128_KB)));
+
+
+The requirements belong to requirement catageries. 
+For instance, ``cub::run_to_run_determinism`` shares category with ``cub::nondeterminism``.
+Only one requirement per category is allowed in the requirements list.
+
+The requirements API allows:
+
+  #. stateful requirements
+  #. providing requirements in any order
+  #. not instantiating unrelated kernels
+  #. failing at compile time if requirements can't be satisfied
+
+
+The requirements API doesn't allow:
+
+  #. multiple requirements from the same category in the requirements list
+  #. providing requirements from the category that is not supported by the algorithm
+
+
+Users do not see passed this API.
+As far as users are concerned, 
+we reserve the right to change default guarantees at any time.
+If there are functional requirements that are critical for users,
+those should be specified specifically. 
+
+All guarantees are ordered within the category.
+We reserve the right to select the implementation with stronger guarantees compared to what user requested.
+The remaining part of this section is only related to CUB developers.
+
+
+Guarantees
+====================================
+
+The best form of documentation is code.
+The default guarantees should be visible to the user in a convenient way outside of the documentation.
+This is achieved as follows:
+
+
+.. code-block:: c++
+
+   struct reduce
+   {
+      using default_guarantees_t =                                                                                          // (1)
+        cub::detail::guarantees::guarantees_t<cub::detail::guarantees::determinism_not_guaranteed_t,
+                                              detail::max_memory_footprint_t>;
+
+      template <class IteratorT, class ReguirementsT = default_guarantees_t>
+      static auto sum(IteratorT begin, IteratorT end, ReguirementsT requirements = default_guarantees_t())                  // (2)
+      {
+        auto guarantees    = cub::detail::requirements::mask(default_guarantees_t(), requirements);                         // (3)
+        auto max_footprint = cub::detail::requirements::get<detail::max_memory_footprint_t>(guarantees);                    // (4)
+        auto determinism = cub::detail::requirements::get<cub::detail::guarantees::run_to_run_deterministic_t>(guarantees); // (5)
```

Comment:
Capturing something that came up in the meeting -- it would be nicer for readability to have a query that doesn't spell out a specific guarantee, but rather the category. Maybe something like

```
enum determinism_t
{
  category, // Always the weakest
  nondeterministic,
  run_to_run_deterministic
};

// ...

namespace category {

// Never exposed outside of detail
using determinism = determinism_holder_t<determinism_t::category>;

}

// ...

auto determinism = cub::detail::requirements::get<cub::detail::category::determinism>(guarantees);
```